### PR TITLE
fix: Allow sync to work with dependencies in source

### PR DESCRIPTION
**Pull request by mildaniel**
*Issue #, if available:*
https://github.com/aws/aws-sam-cli/issues/3707

*Description of changes:*
- Remove node_modules before source-artifact comparison so that node_modules can be packaged in an auto-dependency layer.
- Create a dependency manager class to reduce some of the code duplication

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by mndeveci — COMMENTED** [review]
Looks good in general, left some comments regarding styling.

File: aws_lambda_builders/actions.py [hunk 1]
```diff
@@ -197,3 +187,36 @@ def execute(self):
                 shutil.rmtree(target_path)
             else:
                 os.remove(target_path)
+
+
+class DependencyManager:
```
[the remark this answers is not in the document]
Comment (mildaniel) [reply aws_lambda_builders/actions.py:192]: Absolutely! Good call out.

**Review by mndeveci — APPROVED** [review]

**Review by qingchm — APPROVED** [review]
LGTM!

**Comment by qingchm** [comment]
Side note: at the review session we confirmed that the app builder changes were run locally against CLI integration tests and passed. Discussed the necessity to add in an integration test on this, we will assign new task points to design an integration test.

**Pull request merged**